polaris/ocean/viz/transect.py

```diff
@@ -0,0 +1,159 @@
+import cmocean  # noqa: blah
+import matplotlib.pyplot as plt
+import numpy as np
+from matplotlib.tri import Triangulation
+from mpas_tools.ocean.transects import (
+    find_transect_levels_and_weights,
+    get_outline_segments,
+    interp_mpas_to_transect_triangles,
+)
+from mpas_tools.viz import mesh_to_triangles
+from mpas_tools.viz.transects import (
+    find_planar_transect_cells_and_weights,
+    find_transect_cells_and_weights,
+    make_triangle_tree,
+)
+
+
+def compute_transect(x, y, ds_3d_mesh, spherical=False):
+    """
+    build a sequence of triangles showing the transect intersecting mpas cells
+
+    Parameters
+    ----------
+    x : xarray.DataArray
+        The x or longitude coordinate of the transect
+
+    y : xarray.DataArray
+        The y or latitude coordinate of the transect
+
+    ds_3d_mesh : xarray.Dataset
+        The MPAS-Ocean mesh to use for plotting
+
+    spherical : bool, optional
+        Whether the x and y coordinates are latitude and longitude in degrees
+
+    Returns
+    -------
+    ds_transect : xarray.Dataset
+        The transect dataset
+    """
+
+    ds_tris = mesh_to_triangles(ds_3d_mesh)
+
+    triangle_tree = make_triangle_tree(ds_tris)
+
+    if spherical:
+        ds_transect = find_transect_cells_and_weights(
+            x, y, ds_tris, ds_3d_mesh, triangle_tree, degrees=True)
+    else:
+        ds_transect = find_planar_transect_cells_and_weights(
+            x, y, ds_tris, ds_3d_mesh, triangle_tree)
+
+    cell_indices = ds_transect.horizCellIndices
+    mask = ds_3d_mesh.maxLevelCell.isel(nCells=cell_indices) > 0
+    ds_transect = ds_transect.isel(nSegments=mask)
+
+    ds_transect = find_transect_levels_and_weights(
+        ds_transect, ds_3d_mesh.layerThickness,
+        ds_3d_mesh.bottomDepth, ds_3d_mesh.maxLevelCell - 1)
```

Comment:
I will need to remind myself about the details but I didn't decide to move to triangles lightly.  It was a huge pain in the ass to make the change.  So I must have had a good reason that I couldn't use quads anymore that I am not remembering now. I'll look into it and keep you posted.  If we can move back to quads, that would be great for a lot of reasons.  One is that labeled contours don't work right on triangles.

Reply:
Looking through the docstrings and such, I don't see why we couldn't use quads.  For example,
https://github.com/MPAS-Dev/MPAS-Tools/blob/master/conda_package/mpas_tools/ocean/transects.py#L40-L46
but I don't see why we couldn't use `flat` or `gouraud` shading to accomplish the same.  

What I recall being tricky was masking out the invalid cells (e.g. below bathymetry).  It's straightforward with flat shading -- you just have NaNs for invalid cell values.  But for gouraud, you have a problem that you could have an invalid cell surrounded by 4 valid nodes (e.g. because you have a local column of high bathymetry).  A way to handle this and still use `pcolormesh` would be to add mid-point values to each transect segment.  This would fix the masking issue for bathymetry and would make the interpolation of field values along the transect slightly more accurate.  But it would mean that there would be vertical lines both at intersections of the transect with cell edges and at the midpoint between cell edges.  This would make the outlines of quads less meaningful.

Reply:
We could use the straightforward approach for flat shading and add an midpoint to the transect segments for gouraud shading (presumably discouraging cell outlines in that case).

Reply:
I can see how this goes at least.

Reply:
It took me more than a week but I got the transect plotting to work on quads.  Showing the edges of the quads still does not have the effect that you were after because the quads do not correspond to MPAS-ocean cells (the cells have to be divided into triangles horizontally and half-levels vertically for interpolation to nodes to work correctly).  However, I have stored segments that can be plotted to get MPAS-Ocean layer interfaces and vertical cell boundaries so you can get the effect that you were after.